models/stdp_nn_pre-centered_connection.h

```diff
@@ -0,0 +1,367 @@
+  }
+
+  double
+  depress_( double w, double kminus )
+  {
+    double norm_w = ( w / Wmax_ ) - ( alpha_ * lambda_ * std::pow( w / Wmax_, mu_minus_ ) * kminus );
+    return norm_w > 0.0 ? norm_w * Wmax_ : 0.0;
+  }
+
+  // data members of each connection
+  double weight_;
+  double tau_plus_;
+  double lambda_;
+  double alpha_;
+  double mu_plus_;
+  double mu_minus_;
+  double Wmax_;
+  double Kplus_;
+
+  double t_lastspike_;
+};
+
+
+/**
+ * Send an event to the receiver of this connection.
+ * \param e The event to send
+ * \param t The thread on which this connection is stored.
+ * \param cp Common properties object, containing the stdp parameters.
+ */
+template < typename targetidentifierT >
+inline void
+STDPNNPreCenteredConnection< targetidentifierT >::send( Event& e, thread t, const CommonSynapseProperties& )
+{
+  // synapse STDP depressing/facilitation dynamics
+  double t_spike = e.get_stamp().get_ms();
+
+  // use accessor functions (inherited from Connection< >) to obtain delay and
+  // target
+  Node* target = get_target( t );
+  double dendritic_delay = get_delay();
+
+  // get spike history in relevant range (t1, t2] from post-synaptic neuron
+  std::deque< histentry >::iterator start;
+  std::deque< histentry >::iterator finish;
+
+  // For a new synapse, t_lastspike_ contains the point in time of the last
+  // spike. So we initially read the
+  // history(t_last_spike - dendritic_delay, ..., T_spike-dendritic_delay]
+  // which increases the access counter for these entries.
+  // At registration, all entries' access counters of
+  // history[0, ..., t_last_spike - dendritic_delay] have been
+  // incremented by Archiving_Node::register_stdp_connection(). See bug #218 for
+  // details.
+  target->get_history( t_lastspike_ - dendritic_delay, t_spike - dendritic_delay, &start, &finish );
+  // If there were no post-synaptic spikes between the current pre-synaptic one
+  // t_spike and the previous pre-synaptic one t_lastspike_, there are no pairs
+  // to account.
+  if ( start != finish )
+  {
+    // facilitation due to the first post-synaptic spike start->t_
+    // since the previous pre-synaptic spike t_lastspike_
+
+    double minus_dt;
+    minus_dt = t_lastspike_ - ( start->t_ + dendritic_delay );
+
+    // get_history() should make sure that
+    // start->t_ > t_lastspike_ - dendritic_delay, i.e. minus_dt < 0
+    assert( minus_dt < -1.0 * kernel().connection_manager.get_stdp_eps() );
+
+    weight_ = facilitate_( weight_, Kplus_ * std::exp( minus_dt / tau_plus_ ) );
+
+    // According to the presynaptic-centered nearest-neighbour scheme,
+    // a postsynaptic spike
+    // (we now know there was at least one between t_lastspike_ and t_spike)
+    // erases the state of the synapse,
+    // and all the preceding presynaptic spikes are forgotten.
+    Kplus_ = 0;
+  }
+
+  // depression due to the latest post-synaptic spike finish->t_
+  // before the current pre-synaptic spike t_spike
+  --finish;
```

Comment:
Is shifting this pointer necessary? I think it is not used below, right?

Reply:
You are right, it is not indeed!